packages/flutter/lib/src/widgets/routes.dart

```diff
@@ -1417,6 +1417,84 @@ abstract class ModalRoute<T> extends TransitionRoute<T> with LocalHistoryRoute<T
     return child;
   }
 
+  /// The [DelegatedTransition] provided to the route below this one in the
+  /// navigation stack.
+  ///
+  /// {@macro flutter.widgets.delegatedTransition}
+  ///
+  /// The [ModalRoute] receiving this transition will set it to their
+  /// [receivedTransition] property.
+  DelegatedTransition? get delegatedTransition => null;
+
+  /// The [DelegatedTransition] received from the route above this one in the
+  /// navigation stack.
+  ///
+  /// {@macro flutter.widgets.delegatedTransition}
+  ///
+  /// `receivedTransition` will use the above route's [delegatedTransition] in
+  /// order to show the right route transition when the above route either enters
+  /// or leaves the navigation stack. If not null, the `receivedTransition` will
+  /// wrap the route content through [buildFlexTransitions].
+  DelegatedTransition? receivedTransition;
+
+  /// A method to trigger setting the [delegatedTransition] for a route.
+  ///
+  /// It is possible for a route to supply a different [delegatedTransition]
+  /// depending on the context of the tree at its location. For example Material's
+  /// [ZoomPageTransitionsBuilder] depends on the [Theme] to conditionally provide
+  /// a non-snapshotted [delegatedTransition].
+  ///
+  /// In some cases, the previous route may ask for the [delegatedTransition] of
+  /// a route before it is built. In these cases, the previous route will call
+  /// [setDelegatedTransition] on the the unbuilt route using its own
+  /// [BuildContext]. Override this method to set the [delegatedTransition] based
+  /// on the provided [BuildContext]. This context will be from the PREVIOUS
+  /// route, so do not use logic that requires the [BuildContext] of the current
+  /// route.
+  void setDelegatedTransition(BuildContext context) {}
```

Comment:
It seems weird that we have this method, shouldn't it be subclass's responsibility to find the right time to set the delegate?

Reply:
The subclasses define the actual logic of it, and it allows others of the same parent class to trigger it on each other. It will also be useful for the modal bottom sheet as it needs to give a different transition based on whether it's the first sheet in the view or not, and it likely won't know that before it's built.

Reply:
Instead of finding trigger point for both subclass or parent class, Can we expose a new API in either Route class or modalRoute class that will be called when context may have updated? from the look of the code, you can add a trigger in here which should catch all the cases I think
https://github.com/flutter/flutter/blob/fea5d1ec039422271c7c220c7d79618c401b96d3/packages/flutter/lib/src/widgets/routes.dart#L996